lib/SIL/SIL.cpp

```diff
@@ -105,3 +105,101 @@ swift::getLinkageForProtocolConformance(const NormalProtocolConformance *C,
       return (definition ? SILLinkage::Public : SILLinkage::PublicExternal);
   }
 }
+
+bool SILModule::isTypeMetadataAccessible(CanType type) {
+  assert(type->isLegalFormalType());
+
+  return !type.findIf([&](CanType type) {
+    // Note that this function returns true if the type is *illegal* to use.
+
+    // Ignore non-nominal types.
+    auto decl = type.getNominalOrBoundGenericNominal();
+    if (!decl)
+      return false;
+
+    // Check whether the declaration is inaccessible from the current context.
+    switch (getDeclLinkage(decl)) {
```

Comment:
This is just a minor observation, but I feel that the FormalLinkage enum adds no value. It's only used in a couple of other places, and we should look at AST's access instead.